src/flags.js

```diff
@@ -0,0 +1,649 @@
+					uid: uid,
+					datetime: timestamp
+				}),
+				async.apply(db.sortedSetAdd.bind(db), 'flags:datetime', timestamp, flagId),	// by time, the default
+				async.apply(db.sortedSetAdd.bind(db), 'flags:byReporter:' + uid, timestamp, flagId),	// by reporter
+				async.apply(db.sortedSetAdd.bind(db), 'flags:byType:' + type, timestamp, flagId),	// by flag type
+				async.apply(db.setObjectField.bind(db), 'flagHash:flagId', [type, id, uid].join(':'), flagId),	// save hash for existence checking
+				async.apply(analytics.increment, 'flags')	// some fancy analytics
+			];
+
+			if (targetUid) {
+				tasks.push(async.apply(db.sortedSetAdd.bind(db), 'flags:byTargetUid:' + targetUid, timestamp, flagId));	// by target uid
+			}
+		
+			async.parallel(tasks, function (err, data) {
+				if (err) {
+					return next(err);
+				}
+
+				if (doHistoryAppend) {
+					Flags.update(flagId, uid, { "state": "open" });
+				}
+
+				next(null, flagId);
+			});
+		},
+		async.apply(Flags.get)
+	], callback);
+	// if (!parseInt(uid, 10) || !reason) {
+	// 	return callback();
+	// }
+
+	// async.waterfall([
+	// 	function (next) {
+	// 		async.parallel({
+	// 			hasFlagged: async.apply(Flags.isFlaggedByUser, post.pid, uid),
+	// 			exists: async.apply(Posts.exists, post.pid)
+	// 		}, next);
+	// 	},
+	// 	function (results, next) {
+	// 		if (!results.exists) {
+	// 			return next(new Error('[[error:no-post]]'));
+	// 		}
+
+	// 		if (results.hasFlagged) {
+	// 			return next(new Error('[[error:already-flagged]]'));
+	// 		}
+
+	// 		var now = Date.now();
+	// 		async.parallel([
+	// 			function (next) {
+	// 				db.sortedSetAdd('posts:flagged', now, post.pid, next);
+	// 			},
+	// 			function (next) {
+	// 				db.sortedSetIncrBy('posts:flags:count', 1, post.pid, next);
+	// 			},
+	// 			function (next) {
+	// 				db.incrObjectField('post:' + post.pid, 'flags', next);
+	// 			},
+	// 			function (next) {
+	// 				db.sortedSetAdd('pid:' + post.pid + ':flag:uids', now, uid, next);
+	// 			},
+	// 			function (next) {
+	// 				db.sortedSetAdd('pid:' + post.pid + ':flag:uid:reason', 0, uid + ':' + reason, next);
+	// 			},
+	// 			function (next) {
+	// 				if (parseInt(post.uid, 10)) {
+	// 					async.parallel([
+	// 						async.apply(db.sortedSetIncrBy, 'users:flags', 1, post.uid),
+	// 						async.apply(db.incrObjectField, 'user:' + post.uid, 'flags'),
+	// 						async.apply(db.sortedSetAdd, 'uid:' + post.uid + ':flag:pids', now, post.pid)
+	// 					], next);
+	// 				} else {
+	// 					next();
+	// 				}
+	// 			}
+	// 		], next);
+	// 	},
+	// 	function (data, next) {
+	// 		openNewFlag(post.pid, uid, next);		// removed, used to just update flag to open state if new flag
+	// 	}
+	// ], function (err) {
+	// 	if (err) {
+	// 		return callback(err);
+	// 	}
+	// 	analytics.increment('flags');
+	// 	callback();
+	// });
+};
+
+Flags.exists = function (type, id, uid, callback) {
+	db.isObjectField('flagHash:flagId', [type, id, uid].join(':'), callback);
```

Comment:
can `flagHash:flagId` grow beyond 16mb? Might be better to store is as zset if it can. `[type, id, uid].join(':')` would be the value and `flagId` would be the score